compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -330,8 +330,21 @@ object Implicits:
       if ownEligible.isEmpty then outerEligible
       else if outerEligible.isEmpty then ownEligible
       else
-        val shadowed = ownEligible.map(_.ref.implicitName).toSet
-        ownEligible ::: outerEligible.filterConserve(cand => !shadowed.contains(cand.ref.implicitName))
+        def filter(xs: List[Candidate], remove: List[Candidate]) =
+          val shadowed = remove.map(_.ref.implicitName).toSet
+          xs.filterConserve(cand => !shadowed.contains(cand.ref.implicitName))
+
+        val outer = outerImplicits.uncheckedNN
+        def isWildcardImport(using Context) = ctx.importInfo.nn.isWildcardImport
+        def preferDefinitions = isImport && !outer.isImport
+        def preferNamedImport = isWildcardImport && !isWildcardImport(using outer.irefCtx)
+
+        if !migrateTo3(using irefCtx) && level == outer.level && (preferDefinitions || preferNamedImport) then
```

Comment:
Why the !migrateTo3 condition? I thought Scala 2 has the same preferences of definitions over imports?

Reply:
To counter the fact that we're flattening it in "level".  Otherwise we end up preferring the candidates in an outer scope, which should have a smaller level, over the ones in an inner scope.

I encountered it in CI, and minimised the problem to `tests/pos/i18183.migration.scala`.  Because it runs under `-source:3.0-migration` then outer `implicit F: Async[F]` "shadows" inner `implicit F: Async[M]`, making it impossible to call `semiflatMap` for lack of a `Async[M]` in scope.

Reply:
I see. Yes that makes sense.